// patronum/interval/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.interval = interval;
var _effector = require("effector");
function interval({
  timeout,
  start,
  stop,
  leading = false,
  trailing = false
}) {
  const tick = (0, _effector.createEvent)({
    name: "tick",
    sid: "u782pl"
  });
  const $isRunning = (0, _effector.createStore)(false, {
    name: "$isRunning",
    sid: "-xegewf"
  });
  const $timeout = toStoreNumber(timeout);
  const $notRunning = $isRunning.map(running => !running);
  const saveTimeout = (0, _effector.createEvent)({
    name: "saveTimeout",
    sid: "ab3q7d"
  });
  const $timeoutId = (0, _effector.createStore)(null, {
    name: "$timeoutId",
    sid: "-aj5rja"
  }).on(saveTimeout, (_, {
    timeoutId
  }) => timeoutId);
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const $rejecter = (0, _effector.createStore)(() => {}, {
    name: "$rejecter",
    sid: "opb0ih"
  }).on(saveTimeout, (_, {
    reject
  }) => reject);
  const timeoutFx = (0, _effector.attach)({
    and: {
      source: {
        timeout: $timeout,
        running: $isRunning
      },
      effect: ({
        timeout,
        running
      }) => {
        if (!running) {
          return Promise.reject();
        }
        return new Promise((resolve, reject) => {
          const timeoutId = setTimeout(resolve, timeout);
          saveTimeout({
            timeoutId,
            reject
          });
        });
      }
    },
    or: {
      name: "timeoutFx",
      sid: "9hil7b"
    }
  });
  const cleanupFx = (0, _effector.attach)({
    and: {
      source: {
        timeoutId: $timeoutId,
        rejecter: $rejecter
      },
      effect: ({
        timeoutId,
        rejecter
      }) => {
        rejecter();
        if (timeoutId) clearTimeout(timeoutId);
      }
    },
    or: {
      name: "cleanupFx",
      sid: "-v02m7d"
    }
  });
  (0, _effector.guard)({
    and: [{
      clock: start,
      source: $timeout,
      filter: $notRunning,
      target: timeoutFx
    }],
    or: {
      sid: "-vop0ii"
    }
  });
  if (leading) {
    const onReady = (0, _effector.guard)({
      and: [{
        clock: start,
        filter: $notRunning
      }],
      or: {
        name: "onReady",
        sid: "-cmw7lh"
      }
    });
    (0, _effector.sample)({
      and: [{
        clock: onReady,
        target: tick
      }],
      or: {
        sid: "-v876ny"
      }
    });
  }
  (0, _effector.sample)({
    and: [{
      clock: start,
      fn: () => true,
      target: $isRunning
    }],
    or: {
      sid: "-v6jsxl"
    }
  });
  (0, _effector.guard)({
    and: [{
      clock: timeoutFx.done,
      source: $timeout,
      filter: $isRunning,
      target: timeoutFx
    }],
    or: {
      sid: "-urpcwu"
    }
  });
  (0, _effector.guard)({
    and: [{
      clock: timeoutFx.done,
      filter: $isRunning,
      target: tick.prepend(() => {
        /* to be sure, nothing passed to tick */
      })
    }],
    or: {
      sid: "-ucb4aq"
    }
  });
  if (stop) {
    if (trailing) {
      (0, _effector.sample)({
        and: [{
          clock: stop,
          target: tick
        }],
        or: {
          sid: "6lcn0o"
        }
      });
    }
    $isRunning.on(stop, () => false);
    (0, _effector.sample)({
      and: [{
        clock: stop,
        target: cleanupFx
      }],
      or: {
        sid: "71ao6f"
      }
    });
  }
  return {
    tick,
    isRunning: $isRunning
  };
}
function toStoreNumber(value) {
  if (_effector.is.store(value, {
    sid: "7icamc"
  })) return value;
  if (typeof value === 'number') {
    return (0, _effector.createStore)(value, {
      and: {
        name: '$timeout'
      },
      sid: "knc3em"
    });
  }
  throw new TypeError(`timeout parameter in interval method should be number or Store. "${typeof value}" was passed`);
}
