// patronum/debounce/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.debounce = debounce;
var _effector = require("effector");
function debounce({
  source,
  timeout,
  target
}) {
  if (!_effector.is.unit(source)) throw new TypeError('source must be unit from effector');
  if (_effector.is.domain(source, {
    sid: "k76nst"
  })) throw new TypeError('source cannot be domain');
  const $timeout = toStoreNumber(timeout);
  const saveTimeoutId = (0, _effector.createEvent)({
    name: "saveTimeoutId",
    sid: "vsg63k"
  });
  const $timeoutId = (0, _effector.createStore)(null, {
    and: {
      serialize: 'ignore'
    },
    name: "$timeoutId",
    sid: "lhz3ie"
  }).on(saveTimeoutId, (_, id) => id);
  const saveReject = (0, _effector.createEvent)({
    name: "saveReject",
    sid: "qddkx3"
  });
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const $rejecter = (0, _effector.createStore)(null, {
    and: {
      serialize: 'ignore'
    },
    name: "$rejecter",
    sid: "-o8m8p2"
  }).on(saveReject, (_, rj) => rj);
  const tick = target !== null && target !== void 0 ? target : (0, _effector.createEvent)({
    name: "tick",
    sid: "a51tok"
  });
  const timerBaseFx = (0, _effector.createEffect)(({
    timeout,
    timeoutId,
    rejectPromise
  }) => {
    if (timeoutId) clearTimeout(timeoutId);
    if (rejectPromise) rejectPromise();
    return new Promise((resolve, reject) => {
      saveReject(reject);
      saveTimeoutId(setTimeout(resolve, timeout));
    });
  }, {
    name: "timerBaseFx",
    sid: "-6dos97"
  });
  const timerFx = (0, _effector.attach)({
    and: {
      source: {
        timeoutId: $timeoutId,
        rejectPromise: $rejecter
      },
      mapParams: (timeout, {
        timeoutId,
        rejectPromise
      }) => {
        return {
          timeout,
          timeoutId,
          rejectPromise
        };
      },
      effect: timerBaseFx
    },
    or: {
      name: "timerFx",
      sid: "ca9mop"
    }
  });
  $rejecter.reset(timerFx.done);
  $timeoutId.reset(timerFx.done);

  // It's ok - nothing will ever start unless source is triggered
  const $payload = (0, _effector.createStore)([], {
    and: {
      serialize: 'ignore'
    },
    name: "$payload",
    sid: "jthvlw"
  }).on(source, (_, payload) => [payload]);
  const $canTick = (0, _effector.createStore)(true, {
    and: {
      serialize: 'ignore'
    },
    name: "$canTick",
    sid: "-rehy9s"
  });
  const triggerTick = (0, _effector.createEvent)({
    name: "triggerTick",
    sid: "31ed0l"
  });
  $canTick.on(triggerTick, () => false).on([tick,
  // debounce timeout should be restarted on timeout change
  $timeout,
  // debounce timeout can be restarted in later ticks
  timerFx], () => true);
  const requestTick = (0, _effector.merge)([source,
  // debounce timeout is restarted on timeout change
  $timeout], {
    name: "requestTick",
    sid: "7xoh8q"
  });
  (0, _effector.guard)({
    and: [{
      clock: requestTick,
      filter: $canTick,
      target: triggerTick
    }],
    or: {
      sid: "-1mcdjn"
    }
  });
  (0, _effector.sample)({
    and: [{
      source: $timeout,
      clock: triggerTick,
      target: timerFx
    }],
    or: {
      sid: "-1j1lzh"
    }
  });
  (0, _effector.sample)({
    and: [{
      source: $payload,
      clock: timerFx.done,
      fn: ([payload]) => payload,
      target: tick
    }],
    or: {
      sid: "-1475yq"
    }
  });
  return tick;
}
function toStoreNumber(value) {
  if (_effector.is.store(value, {
    sid: "-mlqvq"
  })) return value;
  if (typeof value === 'number') {
    if (value < 0 || !Number.isFinite(value)) throw new Error(`timeout must be positive number or zero. Received: "${value}"`);
    return (0, _effector.createStore)(value, {
      and: {
        name: '$timeout'
      },
      sid: "-6oihcx"
    });
  }
  throw new TypeError(`timeout parameter in interval method should be number or Store. "${typeof value}" was passed`);
}
