// patronum/throttle/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.throttle = throttle;
var _effector = require("effector");
function throttle({
  source,
  timeout,
  target = (0, _effector.createEvent)({
    name: "target",
    sid: "r328r6"
  })
}) {
  if (!_effector.is.unit(source)) throw new TypeError('source must be unit from effector');
  const $timeout = toStoreNumber(timeout);
  const timerFx = (0, _effector.createEffect)(timeout => new Promise(resolve => setTimeout(resolve, timeout)), {
    name: "timerFx",
    sid: "-pyjx8k"
  });

  // It's ok - nothing will ever start unless source is triggered
  const $payload = (0, _effector.createStore)(null, {
    and: {
      serialize: 'ignore'
    },
    name: "$payload",
    sid: "hinjft"
  }).on(source, (_, payload) => payload);
  const triggerTick = (0, _effector.createEvent)({
    name: "triggerTick",
    sid: "-sf8sou"
  });
  const $canTick = (0, _effector.createStore)(true, {
    and: {
      serialize: 'ignore'
    },
    name: "$canTick",
    sid: "-m0cqc"
  }).on(triggerTick, () => false).on(target, () => true);
  (0, _effector.guard)({
    and: [{
      clock: source,
      filter: $canTick,
      target: triggerTick
    }],
    or: {
      sid: "-hzhj9b"
    }
  });
  (0, _effector.sample)({
    and: [{
      source: $timeout,
      clock: triggerTick,
      target: timerFx
    }],
    or: {
      sid: "-hkn38k"
    }
  });
  (0, _effector.sample)({
    and: [{
      source: $payload,
      clock: timerFx.done,
      target
    }],
    or: {
      sid: "-hhcboe"
    }
  });
  return target;
}
function toStoreNumber(value) {
  if (_effector.is.store(value, {
    sid: "-h0ap6r"
  })) return value;
  if (typeof value === 'number') {
    if (value < 0 || !Number.isFinite(value)) throw new Error(`timeout must be positive number or zero. Received: "${value}"`);
    return (0, _effector.createStore)(value, {
      and: {
        name: '$timeout'
      },
      sid: "-h97n10"
    });
  }
  throw new TypeError(`timeout parameter should be number or Store. "${typeof value}" was passed`);
}
